lib/ABI/FunctionType.cpp

```diff
@@ -52,74 +55,95 @@ bool verify(const SortedVector<RegisterType> &UsedRegisters,
+///
+/// \return `true` if the ABI is valid, `false` otherwise.
+template<model::ABI::Values ABI>
+static constexpr bool verifyReturnValueLocationHelper() {
+  using AT = abi::Trait<ABI>;
+  constexpr auto &LocationRegister = AT::ReturnValueLocationRegister;
 
-static std::optional<model::QualifiedType>
-buildDoubleType(model::Register::Values UpperRegister,
-                model::Register::Values LowerRegister,
-                model::PrimitiveTypeKind::Values CustomKind,
-                model::Binary &TheBinary) {
-  model::PrimitiveTypeKind::Values UpperKind = selectTypeKind(UpperRegister);
-  model::PrimitiveTypeKind::Values LowerKind = selectTypeKind(LowerRegister);
-  if (UpperKind != LowerKind)
-    return std::nullopt;
+  // Skip ABIs that do not allow returning big values.
+  // They do not benefit from this check.
+  if constexpr (LocationRegister != model::Register::Invalid) {
+    if constexpr (isVectorRegisterHelper(LocationRegister)) {
+      // Vector register used as the return value locations are not supported.
+      return false;
+    } else if constexpr (constexprIsContained(AT::CalleeSavedRegisters,
+                                              LocationRegister)) {
+      // Using callee saved register as a return value location doesn't make
+      // much sense: filter those out.
+        return false;
+    } else {
+      // The return value location register can optionally also be the first GPRs,
+      // but only the first one.
+      constexpr auto &GPRs = AT::GeneralPurposeArgumentRegisters;
+      constexpr auto Iterator = constexprFind(GPRs, LocationRegister);
+      if constexpr (Iterator != GPRs.end() && Iterator != GPRs.begin())
+        return false;
+    }
+  }
 
-  size_t UpperSize = model::Register::getSize(UpperRegister);
-  size_t LowerSize = model::Register::getSize(LowerRegister);
-  return model::QualifiedType(TheBinary.getPrimitiveType(CustomKind,
-                                                         UpperSize + LowerSize),
-                              {});
+  return true;
 }
 
-static model::QualifiedType getTypeOrDefault(const model::QualifiedType &Type,
-                                             model::Register::Values Register,
-                                             model::Binary &Binary) {
-  if (Type.UnqualifiedType.get() != nullptr)
-    return Type;
-  else
-    return buildType(Register, Binary);
-}
+/// \brief Replace all the references to `OldKey` with the references to
+///        the newly added `NewType`. Also erases the original type.
+///
+/// \param OldKey The type references of which should be replaced.
+/// \param NewType The new type to replace the references to. It is added to
+///        the model implicitly, no need to do so before calling this function.
+/// \param Model The tuple tree where replacement should take place in.
+///
+/// \return The new path to the added type.
+static model::TypePath
+fullTypeReplacement(const model::Type::Key &OldKey,
```

Comment:
Can we call this `replaceAllUsesWith`? LLVMism ftw.

Reply:
Sure.

Reply:
But on the second thought, maybe the name should also reflect that fact that the type itself also gets replaced? Maybe something like `replaceTypeAndUsesWith`?

Reply:
I'd delegate that to the caller too.
This way this function does one simple and clear thing.
If you prefer we can `replaceTypeAndAllUsesWith` built on top of that, but dropping things is something "serious", forcing the user to do it explicitly is not so bad.
Up to you.

Reply:
I feel like writing stuff like
```
first add this type in there,
then call this function to replace references
and only then remove the old type
```
in the docs is kind of painful and error prone when an alternative option is to have a function that just does all of that without the giving the caller a chance to mess the order up. Not a big deal though.

Reply:
We can have an additional function that removes the old type. But passing around `UpcastablePointer<model::Type> &&` makes me sad.
When you create a `Type`, you should insert it in the model right away. `model::Types` should never live off-model.
In fact, I'd be happy if that was the only way to produced a `model::Type`, but maybe that's too strict.

Reply:
Thinking about this, why don't we still have something like
```cpp
// model::Binary Model
model::StructType &S = Model.create<model::StructType>();
```
for this?

It sounds like a neat-ish way to enforce what you are saying by providing a default way to do it.
Since now the only way to make a new type is to make it manually (`UpcastablePointer::make`) and then record it in, the belief that types can live outside of the model is hard to root out.

What are your thoughts? Should I add a helper like that (as in combine `make` and `record` together)?

Reply:
Yes! `Binary::createType<...>`!

Reply:
1. Renamed it into `replaceAllUsesWith`.
2. Added it 🎉